photon-server/src/main/java/org/photonvision/common/hardware/GPIO/pi/PigpioPin.java

```diff
@@ -0,0 +1,132 @@
+
+public class PigpioPin extends GPIOBase {
+
+    public static final Logger logger = new Logger(PigpioPin.class, LogGroup.General);
+    private static final PigpioSocket piSocket = new PigpioSocket();
+
+    private final boolean isHardwarePWMPin;
+    private final int pinNo;
+
+    private final ArrayList<PigpioPulse> pulses = new ArrayList<>();
+
+    public PigpioPin(int pinNo) {
+        isHardwarePWMPin = pinNo == 12 || pinNo == 13 || pinNo == 17 || pinNo == 18;
+        this.pinNo = pinNo;
+    }
+
+    @Override
+    public int getPinNumber() {
+        return pinNo;
+    }
+
+    @Override
+    protected void setStateImpl(boolean state) {
+        try {
+            piSocket.gpioWrite(pinNo, state);
+        } catch (PigpioException e) {
+            logger.error("gpioWrite FAIL - " + e.getMessage());
+        }
+    }
+
+    @Override
+    public boolean shutdown() {
+        setState(false);
+        return true;
+    }
+
+    @Override
+    public boolean getStateImpl() {
+        try {
+            return piSocket.gpioRead(pinNo);
+        } catch (PigpioException e) {
+            logger.error("gpioWrite FAIL - " + e.getMessage());
+            return false;
+        }
+    }
+
+    @Override
+    protected void blinkImpl(int pulseTimeMillis, int blinks) {
+        boolean repeat = blinks == -1;
+
+        if (repeat) {
+            blinks = 1;
+        }
+
+        try {
+            piSocket.waveTxStop();
+            pulses.clear();
+
+            var startPulse = new PigpioPulse(pinNo, 0, pulseTimeMillis * 1000);
+            var endPulse = new PigpioPulse(0, pinNo, pulseTimeMillis * 1000);
+
+            for (int i = 0; i < blinks; i++) {
+                pulses.add(startPulse);
+                pulses.add(endPulse);
+            }
+
+            piSocket.waveAddGeneric(pulses);
+            var waveId = piSocket.waveCreate();
+
+            if (waveId >= 0) {
+                if (repeat) piSocket.waveSendRepeat(waveId);
+                else piSocket.waveSendOnce(waveId);
+            } else {
+                String error = "";
+                switch (waveId) {
+                    case PI_EMPTY_WAVEFORM:
+                        error = "Waveform empty";
+                        break;
+                    case PI_TOO_MANY_CBS:
+                        error = "Too many CBS";
+                        break;
+                    case PI_TOO_MANY_OOL:
+                        error = "Too many OOL";
+                        break;
+                    case PI_NO_WAVEFORM_ID:
+                        error = "No waveform ID";
+                        break;
+                }
+                logger.error("Failed to send wave: " + error);
+            }
+
+        } catch (PigpioException e) {
+            logger.error("Could not set blink - " + e.getMessage());
+        }
+    }
+
+    @Override
+    protected void setBrightnessImpl(int brightness) {
+        if (isHardwarePWMPin) {
```

Comment:
Should we print an error if the current pin is not a hardware PWM pin?

Reply:
Maybe once. I can also re-add softPWM and attempt using that, but I'd rather just stick to hardware PWM.

Reply:
Resolved in 4cf21c8